collect_app/src/main/java/org/odk/collect/android/openrosa/OpenRosaFormSource.java

```diff
@@ -100,13 +100,27 @@ public ManifestFile fetchManifest(String manifestURL) throws FormSourceException
     }
 
     @Override
-    public @NotNull InputStream fetchForm(String formURL) throws FormSourceException {
-        return fetchFile(formURL);
+    @NotNull
+    public InputStream fetchForm(String formURL) throws FormSourceException {
+        InputStream formFile = mapException(() -> openRosaXMLFetcher.getFile(formURL, null));
+
+        if (formFile != null) {
+            return formFile;
+        } else {
+            throw new FormSourceException(FETCH_ERROR);
+        }
     }
 
     @Override
+    @NotNull
     public InputStream fetchMediaFile(String mediaFileURL) throws FormSourceException {
-        return mapException(() -> openRosaXMLFetcher.getFile(mediaFileURL, null));
+        InputStream mediaFile = mapException(() -> openRosaXMLFetcher.getFile(mediaFileURL, null));
+
+        if (mediaFile != null) {
```

Comment:
Wouldn't it be better to check this in mapException and throw `new FormSourceException(FETCH_ERROR)` there? Then it would be in one place.

Reply:
Yeah, I think you're right. It feels like a good place for some defensive programming.